src/Adapters/Auth/oauth2.js

```diff
@@ -0,0 +1,231 @@
+      const appidField = options.appidField;
+      if (response && response[appidField]) {
+        const responseValue = response[appidField];
+        if (Array.isArray(responseValue)) {
+          for (var idx = responseValue.length - 1; idx >= 0; idx--) {
+            if (appIds.indexOf(responseValue[idx]) != -1) {
+              return Promise.resolve();
+            }
+          }
+        } else {
+          if (appIds.indexOf(responseValue) != -1) {
+            return Promise.resolve();
+          }
+        }
+      }
+      return Promise.reject(new Parse.Error(
+        Parse.Error.OBJECT_NOT_FOUND,
+        'OAuth2: the access_token\'s appID is empty or is not in the list of permitted appIDs in the auth configuration.'));
+    });
+  } else {
+    return Promise.resolve();
+  }
+}
+
+function _isLoggingEnabled(options) {
+  return options && options.debug;
+}
+
+function _objectToString(object) {
+  return JSON.stringify(object, null, 2);
+}
+
+function _getLogger(functionName) {
+  const prefix = scriptName + '.' + functionName + '(): ';
+  var getFormat = function(arg) {
+    return (new Date()).toISOString() + ' ' + prefix + arg;
+  }
+  return {
+    info: function() {
+      arguments[0] = getFormat(arguments[0]);
+      return logger.info.apply(logger, arguments);
+    },
+    error: function () {
+      arguments[0] = getFormat(arguments[0]);
+      return logger.error.apply(logger, arguments);
+    },
+    warn: function() {
+      arguments[0] = getFormat(arguments[0]);
+      return logger.warn.apply(logger, arguments);
+    },
+    verbose: function() {
+      arguments[0] = getFormat(arguments[0]);
+      return logger.verbose.apply(logger, arguments);
+    },
+    debug: function() {
+      arguments[0] = getFormat(arguments[0]);
+      return logger.debug.apply(logger, arguments);
+    },
+    silly: function() {
+      arguments[0] = getFormat(arguments[0]);
+      return logger.silly.apply(logger, arguments);
+    }
+  };
+}
+
+// A promise wrapper for api requests
+function requestJson(options, access_token, loggingEnabled) {
+  const _logger = _getLogger(requestJson.name);
+  if (loggingEnabled) {
+    _logger.verbose('options = %s', _objectToString(options));
+  }
+  return new Promise(function(resolve, reject) {
+    if (!options || !options.tokenIntrospectionEndpointUrl) {
+      reject(new Parse.Error(Parse.Error.OBJECT_NOT_FOUND, 'OAuth2 token introspection endpoint URL is missing from configuration!'));
+    }
+    const url = options.tokenIntrospectionEndpointUrl;
+    if (loggingEnabled) {
+      _logger.verbose('url = %s', url);
+    }
+    const parsedUrl = Url.parse(url);
+    const postData = Querystring.stringify({
+      'token': access_token
+    });
+    const headers = {
+      'Content-Type': 'application/x-www-form-urlencoded',
+      'Content-Length': Buffer.byteLength(postData)
+    }
+    // Note: the "authorizationHeader" adapter config must contain the raw value.
```

Comment:
this should be above in the main summary doc that you've written probably.

Also the comment is wrong:

```
// Thus if HTTP Basic authorization is to be used, it must contain the
//   base64 encoded version of the concatenated <username> + ":" + <password> string.
```

As it also requries `Basic` to be written as the 1st part.

Reply:
You're right. Fixed.